src/query/shorthand.ts

```diff
@@ -277,3 +282,205 @@ export function fieldDef(encQ: EncodingQuery,
   }
   return fieldAndParams;
 }
+
+const CHANNEL_INDEX = toMap(CHANNELS);
+const AGGREGATE_OP_INDEX = toMap(AGGREGATE_OPS);
+const SINGLE_TIMEUNIT_INDEX = toMap(SINGLE_TIMEUNITS);
+const MULTI_TIMEUNIT_INDEX = toMap(MULTI_TIMEUNITS);
+
+export function parse(shorthand: string): SpecQuery {
+  // TODO(https://github.com/uwdata/compassql/issues/259):
+  // Do not split directly, but use an upgraded version of `getClosingBraceIndex()`
+  let splitShorthand = shorthand.split('|');
+
+  let specQ: SpecQuery = {mark: splitShorthand[0], encodings: [] as EncodingQuery[]};
+
+  for (let i = 1; i < splitShorthand.length; i++) {
+    let part = splitShorthand[i];
+    const splitPart = splitWithTail(part, ':', 1);
+    const splitPartKey = splitPart[0];
+    const splitPartValue = splitPart[1];
+
+    if (CHANNEL_INDEX[splitPartKey] || splitPartKey === '?') {
+      shorthandParser.encoding(specQ, splitPartKey, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'calculate') {
+      shorthandParser.calculate(specQ, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'filter') {
+      shorthandParser.filter(specQ, splitPartValue);
+      continue;
+    }
+
+    if (splitPartKey === 'filterInvalid') {
+      shorthandParser.filterInvalid(specQ, splitPartValue);
+      continue;
+    }
+  }
+
+  return specQ;
+}
+
+/**
+ * Split a string n times into substrings with the specified delimiter and return them as an array.
+ * @param str The string to be split
+ * @param delim The delimiter string used to separate the string
+ * @param number The value used to determine how many times the string is split
+ */
+export function splitWithTail(str: string, delim: string, count: number) {
+  let parts = str.split(delim);
```

Comment:
I think this is unnecessary expensive, especially when you use it with fieldDef with params. 
(Imagine how many `,` could there be inside your stringified JSON. 

Why don't you just call `indexOf` with correct startIndex for `count` times, each time extract the correct substring.  Then extract the last portion of substring at the end.  That will be way more efficient.